### ISSUE TITLE
error: ambiguous overload for ‘operator*’  in test_list.cpp

### ISSUE THREAD

**Issue by Kevin-Delnoije** [NVIDIA/cudnn-frontend#47]
hey i have cudatoolkit 12.2 and cudnn 8.9 installed on ubuntu 22.04 lts
when following the samples build instructions below i encounter the error in the title
```bash
 mkdir build; cd build
cmake -DCUDNN_PATH=/usr/lib/x86_64-linux-gnu/ -DCUDAToolkit_ROOT=/usr/local/cuda/bin ..
 cmake --build . -j8
```
the configure step seems fine
<details>
  <summary>cmake configure output</summary>

```bash
-- The C compiler identification is GNU 11.3.0
-- The CXX compiler identification is GNU 11.3.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found CUDAToolkit: /usr/local/cuda/include (found version "12.2.91") 
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- cudnn found at /usr/lib/x86_64-linux-gnu/libcudnn.so.
-- cudnn_adv_infer found at /usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.
-- cudnn_adv_train found at /usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.
-- cudnn_cnn_infer found at /usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.
-- cudnn_cnn_train found at /usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.
-- cudnn_ops_infer found at /usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.
-- cudnn_ops_train found at /usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.
-- Found LIBRARY: /usr/include  
-- cuDNN: /usr/lib/x86_64-linux-gnu/libcudnn.so
-- cuDNN: /usr/include
-- Configuring done
-- Generating done
-- Build files have been written to: /home/kevin/code/thirdparty/cudnn-frontend/build
```

</details>

then in test_list.cpp there is an overload resolution error of operator * because of the datatypes
<details>
  <summary>nvcc error</summary>

```bash
/home/kevin/code/thirdparty/cudnn-frontend/samples/test_list.cpp: In function ‘void CATCH2_INTERNAL_TEST_102()’:
/home/kevin/code/thirdparty/cudnn-frontend/samples/test_list.cpp:2127:50: error: ambiguous overload for ‘operator*’ (operand types are ‘half’ and ‘float’)
 2127 |         half globalScaleOutput = afterConvOutput * scale1.hostPtr[0] * scale2.hostPtr[0];
      |                                  ~~~~~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~
      |                                  |                                 |
      |                                  half                              float
/home/kevin/code/thirdparty/cudnn-frontend/samples/test_list.cpp:2127:50: note: candidate: ‘operator*(int, float)’ (built-in)
 2127 |         half globalScaleOutput = afterConvOutput * scale1.hostPtr[0] * scale2.hostPtr[0];
      |                                  ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
/home/kevin/code/thirdparty/cudnn-frontend/samples/test_list.cpp:2127:50: note: candidate: ‘operator*(long long unsigned int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/test_list.cpp:2127:50: note: candidate: ‘operator*(long long int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/test_list.cpp:2127:50: note: candidate: ‘operator*(long unsigned int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/test_list.cpp:2127:50: note: candidate: ‘operator*(long int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/test_list.cpp:2127:50: note: candidate: ‘operator*(unsigned int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/test_list.cpp:2127:50: note: candidate: ‘operator*(float, float)’ (built-in)
In file included from /usr/local/cuda/include/cuda_fp16.h:4070,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/fp16_emu.h:28,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/fp16_dev.h:27,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/helpers.h:36,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:25,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/test_list.cpp:27:
/usr/local/cuda/include/cuda_fp16.hpp:618:44: note: candidate: ‘__half operator*(const __half&, const __half&)’
  618 | __CUDA_HOSTDEVICE__ __forceinline__ __half operator*(const __half &lh, const __half &rh) { return __hmul(lh, rh); }
      |                                            ^~~~~~~~
In file included from /home/kevin/code/thirdparty/cudnn-frontend/samples/test_list.cpp:27:
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h: In instantiation of ‘void scale_and_bias_tensor_cpu(const inputType*, outputType*, const scale_bias_type*, const scale_bias_type*, int64_t, const int64_t*) [with inputType = __half; scale_bias_type = float; outputType = float; int64_t = long int]’:
/home/kevin/code/thirdparty/cudnn-frontend/samples/test_list.cpp:1560:50:   required from here
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:128:38: error: ambiguous overload for ‘operator*’ (operand types are ‘const __half’ and ‘const float’)
  128 |         outputData[i] = inputData[i] * scaleData[c] + biasData[c];
      |                         ~~~~~~~~~~~~~^~~~~~~~~~~~
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:128:38: note: candidate: ‘operator*(int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:128:38: note: candidate: ‘operator*(long long unsigned int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:128:38: note: candidate: ‘operator*(long long int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:128:38: note: candidate: ‘operator*(long unsigned int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:128:38: note: candidate: ‘operator*(long int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:128:38: note: candidate: ‘operator*(unsigned int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:128:38: note: candidate: ‘operator*(float, float)’ (built-in)
In file included from /usr/local/cuda/include/cuda_fp16.h:4070,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/fp16_emu.h:28,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/fp16_dev.h:27,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/helpers.h:36,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:25,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/test_list.cpp:27:
/usr/local/cuda/include/cuda_fp16.hpp:618:44: note: candidate: ‘__half operator*(const __half&, const __half&)’
  618 | __CUDA_HOSTDEVICE__ __forceinline__ __half operator*(const __half &lh, const __half &rh) { return __hmul(lh, rh); }
      |                                            ^~~~~~~~
In file included from /home/kevin/code/thirdparty/cudnn-frontend/samples/test_list.cpp:27:
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h: In instantiation of ‘void gen_stats_cpu(const T_ELEM*, std::vector<std::pair<float, float> >&, int64_t, const int64_t*) [with T_ELEM = __half; int64_t = long int]’:
/home/kevin/code/thirdparty/cudnn-frontend/samples/test_list.cpp:1650:24:   required from here
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:169:75: error: ambiguous overload for ‘operator+’ (operand types are ‘float’ and ‘const __half’)
  169 |         outputData[channel_index].first = outputData[channel_index].first + inputData[i];
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:169:75: note: candidate: ‘operator+(float, int)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:169:75: note: candidate: ‘operator+(float, long long unsigned int)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:169:75: note: candidate: ‘operator+(float, long long int)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:169:75: note: candidate: ‘operator+(float, long unsigned int)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:169:75: note: candidate: ‘operator+(float, long int)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:169:75: note: candidate: ‘operator+(float, unsigned int)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:169:75: note: candidate: ‘operator+(float, float)’ (built-in)
In file included from /usr/local/cuda/include/cuda_fp16.h:4070,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/fp16_emu.h:28,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/fp16_dev.h:27,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/helpers.h:36,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:25,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/test_list.cpp:27:
/usr/local/cuda/include/cuda_fp16.hpp:606:44: note: candidate: ‘__half operator+(const __half&, const __half&)’
  606 | __CUDA_HOSTDEVICE__ __forceinline__ __half operator+(const __half &lh, const __half &rh) { return __hadd(lh, rh); }
      |                                            ^~~~~~~~
In file included from /home/kevin/code/thirdparty/cudnn-frontend/samples/test_list.cpp:27:
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:182:37: error: ambiguous overload for ‘operator-’ (operand types are ‘const __half’ and ‘float’)
  182 |         T_ELEM diff = (inputData[i] - outputData[channel_index].first) * (inputData[i] - outputData[channel_index].first);
      |                       ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:182:37: note: candidate: ‘operator-(int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:182:37: note: candidate: ‘operator-(long long unsigned int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:182:37: note: candidate: ‘operator-(long long int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:182:37: note: candidate: ‘operator-(long unsigned int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:182:37: note: candidate: ‘operator-(long int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:182:37: note: candidate: ‘operator-(unsigned int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:182:37: note: candidate: ‘operator-(float, float)’ (built-in)
In file included from /usr/local/cuda/include/cuda_fp16.h:4070,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/fp16_emu.h:28,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/fp16_dev.h:27,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/helpers.h:36,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:25,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/test_list.cpp:27:
/usr/local/cuda/include/cuda_fp16.hpp:612:44: note: candidate: ‘__half operator-(const __half&, const __half&)’
  612 | __CUDA_HOSTDEVICE__ __forceinline__ __half operator-(const __half &lh, const __half &rh) { return __hsub(lh, rh); }
      |                                            ^~~~~~~~
In file included from /home/kevin/code/thirdparty/cudnn-frontend/samples/test_list.cpp:27:
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:182:88: error: ambiguous overload for ‘operator-’ (operand types are ‘const __half’ and ‘float’)
  182 |         T_ELEM diff = (inputData[i] - outputData[channel_index].first) * (inputData[i] - outputData[channel_index].first);
      |                                                                          ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:182:88: note: candidate: ‘operator-(int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:182:88: note: candidate: ‘operator-(long long unsigned int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:182:88: note: candidate: ‘operator-(long long int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:182:88: note: candidate: ‘operator-(long unsigned int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:182:88: note: candidate: ‘operator-(long int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:182:88: note: candidate: ‘operator-(unsigned int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:182:88: note: candidate: ‘operator-(float, float)’ (built-in)
In file included from /usr/local/cuda/include/cuda_fp16.h:4070,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/fp16_emu.h:28,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/fp16_dev.h:27,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/helpers.h:36,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:25,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/test_list.cpp:27:
/usr/local/cuda/include/cuda_fp16.hpp:612:44: note: candidate: ‘__half operator-(const __half&, const __half&)’
  612 | __CUDA_HOSTDEVICE__ __forceinline__ __half operator-(const __half &lh, const __half &rh) { return __hsub(lh, rh); }
      |                                            ^~~~~~~~
In file included from /home/kevin/code/thirdparty/cudnn-frontend/samples/test_list.cpp:27:
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:183:77: error: ambiguous overload for ‘operator+’ (operand types are ‘float’ and ‘__half’)
  183 |         outputData[channel_index].second = outputData[channel_index].second + diff;
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:183:77: note: candidate: ‘operator+(float, int)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:183:77: note: candidate: ‘operator+(float, long long unsigned int)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:183:77: note: candidate: ‘operator+(float, long long int)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:183:77: note: candidate: ‘operator+(float, long unsigned int)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:183:77: note: candidate: ‘operator+(float, long int)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:183:77: note: candidate: ‘operator+(float, unsigned int)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:183:77: note: candidate: ‘operator+(float, float)’ (built-in)
In file included from /usr/local/cuda/include/cuda_fp16.h:4070,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/fp16_emu.h:28,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/fp16_dev.h:27,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/helpers.h:36,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:25,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/test_list.cpp:27:
/usr/local/cuda/include/cuda_fp16.hpp:606:44: note: candidate: ‘__half operator+(const __half&, const __half&)’
  606 | __CUDA_HOSTDEVICE__ __forceinline__ __half operator+(const __half &lh, const __half &rh) { return __hadd(lh, rh); }
      |                                            ^~~~~~~~
In file included from /home/kevin/code/thirdparty/cudnn-frontend/samples/test_list.cpp:27:
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h: In instantiation of ‘void batch_normalize(const T_ELEM*, T_ELEM*, const std::vector<std::pair<float, float> >&, int64_t, const int64_t*) [with T_ELEM = __half; int64_t = long int]’:
/home/kevin/code/thirdparty/cudnn-frontend/samples/test_list.cpp:1652:26:   required from here
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:204:39: error: ambiguous overload for ‘operator-’ (operand types are ‘const __half’ and ‘const float’)
  204 |         outputData[i] = (inputData[i] - stats[batch_index].first) / (T_ELEM) std::sqrt(stats[batch_index].second);
      |                         ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:204:39: note: candidate: ‘operator-(int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:204:39: note: candidate: ‘operator-(long long unsigned int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:204:39: note: candidate: ‘operator-(long long int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:204:39: note: candidate: ‘operator-(long unsigned int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:204:39: note: candidate: ‘operator-(long int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:204:39: note: candidate: ‘operator-(unsigned int, float)’ (built-in)
/home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:204:39: note: candidate: ‘operator-(float, float)’ (built-in)
In file included from /usr/local/cuda/include/cuda_fp16.h:4070,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/fp16_emu.h:28,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/fp16_dev.h:27,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/helpers.h:36,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/cpu_references.h:25,
                 from /home/kevin/code/thirdparty/cudnn-frontend/samples/test_list.cpp:27:
/usr/local/cuda/include/cuda_fp16.hpp:612:44: note: candidate: ‘__half operator-(const __half&, const __half&)’
  612 | __CUDA_HOSTDEVICE__ __forceinline__ __half operator-(const __half &lh, const __half &rh) { return __hsub(lh, rh); }
      |                                            ^~~~~~~~
g
```

</details>

**Comment by Anerudhan** [comment]
Thanks for raising this.

This has been fixed in v0.9.2 release.
The root cause of the regression, is a nvcc update to how it handles fp16 type.

Let me know if you are still seeing the issue.

**Comment by Kevin-Delnoije** [comment]
the issue is solved! thanks for the quick fix
